README
NAME
    CGI::Inspect - Inspect and debug CGI apps with an in-browser UI

SYNOPSIS
      use CGI::Inspect;

      print "Content-type: text/html\n\n";

      my $food = "toast";

      for my $i (1..10) {
        print "$i cookies for me to eat...<br>";
        inspect() if $i == 5; # be sure to edit $toast :)
      }

      print "I also like $food!";

DESCRIPTION
    This class is a drop-in web based inspector for plain CGI (or CGI-based)
    applications. Include the library, and then call inspect(). In your
    server error logs you'll see something like "Please connect to
    localhost:8080". When you do, you'll be greeted with an inspection UI
    which includes a stack trace, REPL, and other goodies.

    To work it's magic this modules uses Continuity, Devel::LexAlias,
    Devel::StackTrace::WithLexicals, and their prerequisites (such as
    PadWalker). Remember that you can always install such things locally for
    debugging -- no need to install them systemwide (in case you are afraid
    of the power that they provide).

EXPORTED SUBS
  inspect()
    This starts the Continuity server and inspector on the configured port
    (defaulting to 8080). You can pass it parameters which it will then pass
    on to CGI::Inspect->new. The most useful one is probably port:

      inspect( port => 12345 );

    Another useful parameter is plugins. The default is:

        plugins => [qw( BasicLook Exit REPL CallStack )]

PLUGINS
    CGI::Inspect comes with a few plugins by default:

    *   REPL - A simple Read-Eval-Print prompt

    *   StackTrace - A pretty stack trace (with lexical editing!)

    *   Exit - Stop inspecting

  Creating Plugins
    Plugins are easy to create! They are really just subroutines that return
    a string for what they want printed. All of the built-in plugins actuall
    inherti from CGI::Inspect::Plugin, which just provides some convenience
    methods. The main CGI::Inspect module will create an instance of your
    plugin with Plugin->new, and then will execute it with $plugin->process.

    Plugins can, however, make use of Continuity, including utilizing
    callbacks. Here is the complete source to the 'Exit' plugin, as a fairly
    simple example.

      package CGI::Inspect::Plugin::Exit;

      use strict;
      use base 'CGI::Inspect::Plugin';

      sub process {
        my ($self) = @_;
        my $exit_link = $self->request->callback_link(
          Exit => sub {
            $self->manager->{do_exit} = 1;
          }
        );
        my $output = qq{
          <div class=dialog title="Exit">
            $exit_link
          </div>
        };
        return $output;
      }

      1;

    For now, read the source of the default plugins for further ideas. And
    of course email the author if you have any questions or ideas!

METHODS
    These methods are all internal. All you have to do is call inspect().

  CGI::Inspect->new()
    Create a new monitor object.

  $self->start_inspecting
    Load plugins and start inspecting!

  $self->start_server
    Initialize the Continuity server, and begin the run loop.

  $self->display
    Display the current page, based on $self->{content}

  $self->request
    Returns the current request obj

  $self->load_plugins
    Load all of our plugins.

  $self->main
    This is executed as the entrypoint for inspector sessions.

SEE ALSO
    Carp::REPL

AUTHOR
      Brock Wilcox <awwaiid@thelackthereof.org> - http://thelackthereof.org/

COPYRIGHT
      Copyright (c) 2008-2009 Brock Wilcox <awwaiid@thelackthereof.org>. All rights
      reserved.  This program is free software; you can redistribute it and/or
      modify it under the same terms as Perl itself.



Changes
Revision history for CGI::Inspect. See Changes-darcs or the main repository for
a detailed list of internal changes.

Version 0.04 (released 2009.06.06)
  * Update documentation
  * Update CallStack plugin, now includes editability
  * Add experimental CGI::Inspect::Here
  * Teach the REPL to deal without lexicals
  * Print plugin load errors to STDERR

Version 0.03 (released 2009.06.04)
  * Add missing files

Version 0.02 (released 2009.05.30)
  * Renamed to CGI::Inspect
  * Significantly refactored to be lighter
  * Only simple REPL and simple StackTrace plugins for now
  * Added DIE handler to auto-start

Version 0.01 (released 2008.10.14)
  * Initial release, just to get it out there
  * Includes JS and image dependencies


Makefile.PL
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME              => 'CGI::Inspect',
    VERSION_FROM      => 'lib/CGI/Inspect.pm',
    ABSTRACT_FROM     => 'lib/CGI/Inspect.pm',
    AUTHOR            => 'Brock Wilcox <awwaiid@thelackthereof.org>',
    LICENSE           => 'perl',
    PREREQ_PM         => {
      'Continuity'                                         => '1.0',
      'PadWalker'                                          => 0,
      'Devel::LexAlias'                                    => 0,
      'Devel::StackTrace::WithLexicals'                    => '0.01',
      # 'Coro::Generator'                                    => 0,
      # 'Method::Signatures'                                 => '20081008',
      # 'Moose'                                              => '0.58',
      # 'Moose::Exporter'                                    => 0,
      # 'Module::Refresh'                                    => 0,
    },
);


lib/CGI/Inspect/Plugin/REPL.pm
package CGI::Inspect::Plugin::REPL;

use strict;
use base 'CGI::Inspect::Plugin';
use Devel::StackTrace::WithLexicals;
use Data::Dumper;

sub new {
  my $class = shift;
  my $self = $class->SUPER::new(@_);
  $self->{trace} = Devel::StackTrace::WithLexicals->new;
  return $self;
}

sub process {
  my $self = shift;
  my $cmd = $self->param('cmd');
  my $output = $self->{output} || '';
  
  my $trace = $self->{trace};
  $trace->reset_pointer;
  my $frame;
  while($frame = $trace->next_frame) {
    last if $frame->package() !~ /^(Continuity|Coro|CGI::Inspect)/;
  }
  use Data::Dumper;
  
  my $package = $frame->package;

  my $declarations = join "\n",
                     map {"my $_;"}
                     keys %{ $frame->lexicals };

  my $aliases = q|
    while (my ($k, $v) = each %{ $frame->lexicals }) {
      Devel::LexAlias::lexalias 0, $k, $v;
    }
  |;

  my $code = qq|
    package $package;
    use Devel::LexAlias;
    no warnings 'misc'; # declaration in same scope masks earlier instance
    no strict 'vars';   # so we get all the global variables in our package
    $declarations
    sub __execute_cmd {
      $aliases
      return $cmd
    }
    return __execute_cmd();
  |;

  if($cmd) {
      local $Data::Dumper::Terse = 1;
      my $new_output = "\n> " . $cmd . "\n";
      $new_output .= Dumper(eval($code));
      $new_output .= "Error: $@\n" if $@;
      $new_output =~ s{<}{\&lt;}g;
      $output .= $new_output;
  }
  $self->{output} = $output;

  $output = qq{
    <div class=dialog id=repl title='Read-Eval-Print'>
      <pre>
        $output
      </pre>
      &gt; <input type=text name=cmd id=cmd size=40>
      <input type=submit name=send value="Send"></pre>
      <script>document.getElementById('cmd').focus()</script>
    </div>
  };

  return $output;
}

1;



lib/CGI/Inspect/Plugin/FileEdit.pm
package CGI::Inspect::Plugin::FileEdit;

use Moose;
use Method::Signatures;
use File::Slurp;
use HTML::Entities;

extends 'MooseX::Continuity::Widget';

method main {
  my $counter = 1;
  while(1) {
    $self->print("<div class=dialog title='File Edit'>");
    my @files = sort values %INC;
    @files = grep !/home/, @files;
    @files = map { my $f = $_; "<li>" . $self->cb_link( $f => sub {
      $self->edit_file($f);
    }) . "</li>" } @files;
    $self->print("<ul>@files</ul>");
    $self->print($self->cb_button('gah'   => sub { print STDERR "Gah\n" }));
    $self->print("</div>");
    $self->yield(1);
  }
}

method edit_file($filename) {
  my $content = read_file($filename);
  $content = encode_entities($content);
  my $textarea_name = $self->field_name('content');
  while(1) {
    my $action = '';
    $self->print(qq|
      <div class=dialog title='File Edit'>
        <textarea style="width: 100%; height: 30em;" name="$textarea_name">$content</textarea>
        | . $self->cb_button('Cancel' => sub { print STDERR "Got cancel\n"; $action = 'cancel' })
        .   $self->cb_button('Save'   => sub { $action = 'save' })
        .   $self->cb_link('gah'   => sub { print STDERR "Gah\n" })
        . '</div>'
    );
    $self->yield(1);
    print STDERR "Action: $action\n\n";
    if($action eq 'save') {
      $content = $self->param($textarea_name);
      write_file($filename);
      require $filename;
    }
    return if $action;
  }
}

1;




lib/CGI/Inspect/Plugin/CallStack.pm
package CGI::Inspect::Plugin::CallStack;

use strict;
use base 'CGI::Inspect::Plugin';
use Devel::StackTrace::WithLexicals;
use Data::Dumper;
use CGI qw( escapeHTML );

sub new {
  my $class = shift;
  my $self = $class->SUPER::new(@_);
  $self->{trace} = Devel::StackTrace::WithLexicals->new();
  return $self;
}

sub print_trace {
  my $self = shift;
  my $output = "<div class=dialog id=stacktrace title='Stacktrace'><ul>";
  my $trace = $self->{trace};
  $trace->reset_pointer;
  my $frame_index = 0;
  my $frame; 
  while($frame = $trace->next_frame) {
    last if $frame->package() !~ /^(Continuity|Coro|CGI::Inspect)/
      || $frame->subroutine() eq 'CGI::Inspect::inspect';
    $frame_index++;
  }
  while($frame) {
    my $next_frame = $trace->next_frame;
    my $subname = $next_frame ? $next_frame->subroutine : '[GLOBAL]';
    $output .= "<li>$subname"
      . " (" . $frame->filename . ":" . $frame->line . ")"
    ;
    $output .= $self->print_lexicals($frame->lexicals, $frame_index);
    $output .= "</li>";
    $frame = $next_frame;
    $frame_index++;
  }
  $output .= "</ul></div>";
  return $output;
}

sub get_field_name {
  return time() . rand();
}

sub print_lexicals {
  my ($self, $lexicals, $frame_index) = @_;
  my $output = '<ul>';
  local $Data::Dumper::Terse = 1;
  local $Data::Dumper::Indent = 1;
    
  foreach my $var (sort keys %$lexicals) {
    my $val;
    if(ref $lexicals->{$var} eq 'REF' || ref $lexicals->{$var} eq 'SCALAR') {
      $val = Dumper(${ $lexicals->{$var} });
    } else {
      $val = Dumper( $lexicals->{$var} );
    }
    chomp $val;
    $val = escapeHTML($val);
    my $edit_link = $self->request->callback_link(
      "$var" => sub {
        my $save_button = $self->request->callback_submit(
          Save => sub {
            my $val = $self->param('blah');
            if(ref $lexicals->{$var} eq 'REF' || ref $lexicals->{$var} eq 'SCALAR') {
              ${ $lexicals->{$var} } = eval($val);
            } elsif(ref $lexicals->{$var} eq 'ARRAY') {
              @{ $lexicals->{$var} } = @{ eval($val) };
            } elsif(ref $lexicals->{$var} eq 'HASH') {
              %{ $lexicals->{$var} } = %{ eval($val) };
            }
          }
        );
        $self->{output} = qq{
          <div class=dialog id=stacktrace title='Stacktrace'>
            $var = <textarea name=blah style="width: 100%; height: 80%">$val</textarea><br>
            $save_button
            <input type=submit name=cancel value="Cancel">
          </div>
        };
      }
    );
    $output .= qq{
      <li>
        <table border=0>
          <tr><td valign=top>$edit_link =</td>
          <td valign=top><pre>$val</pre></td>
        </table>
      </li>};
  }

  $output .= "</ul>";
  return $output;
}


sub process {
  my ($self) = @_;
  if($self->{output}) {
    my $output = $self->{output};
    $self->{output} = undef;
    return $output;
  }
  return $self->print_trace;
}

1;



lib/CGI/Inspect/Plugin/AdvancedREPL.pm
package CGI::Inspect::Plugin::REPL;

our $VERSION = '0.01';

=head1 NAME

Continuity::Monitor::REPL - Web based REPL

=head1 SYNOPSYS

  use strict;
  use Continuity;
  use Continuity::REPL;

  use vars qw( $repl $server );

  $repl = Continuity::REPL->new;
  $server = Continuity->new( port => 8080 );
  $server->loop;

  sub main {
    my $request = shift;
    my $count = 0;
    while(1) {
      $count++;
      $print("Count: $count");
      $request->next;
    }
  }

The command line interaction looks like this:

  main:001:0> $server
  $Continuity1 = Continuity=HASH(0x86468c8);

  main:002:0> $server->{mapper}->{sessions}
  $HASH1 = {
             19392613106888830468 => Coro::Channel=ARRAY(0x8d82038),
             58979072056380208100 => Coro::Channel=ARRAY(0x8d78890)
           };

  main:003:0> Coro::State::list()                                                
  $ARRAY1 = [
              Coro=HASH(0x8d82208),
              Coro=HASH(0x8d78aa0),
              Coro=HASH(0x8d38b98),
              Coro=HASH(0x8d38a38),
              Coro=HASH(0x8b99248),
              Coro=HASH(0x825d6c8),
              Coro=HASH(0x81d7568),
              Coro=HASH(0x81d7518),
              Coro=HASH(0x81d7448)
            ];

=head1 DESCRIPTION

This provides a Devel::REPL shell for Continuity applications.

For now it is just amusing, but it will become useful once it can run the shell
within the context of individual sessions. Then it might be a nice diagnostic
or perhaps even development tool. Heck... maybe we can throw in a web interface
to it...

Also, this library forces the PERL_RL environment variable to 'Perl' since I
haven't been able to figure out how to hack Term::ReadLine::Gnu yet.

=cut

{
  package Devel::REPL::Continuity;
  use Moose;
  use Method::Signatures;
  extends 'Devel::REPL';
  with qw(
    MooseX::Coro
    MooseX::Continuity::CallbackLinks
  );

  has request => ( is => 'rw' );
  has output_result => (is => 'rw', default => '');

  has 'term' => (
    is => 'rw', required => 1,
    default => sub { }
  );

  no warnings 'redefine';

  sub print {
    my ($self, @ret) = @_;
    my $out = "@ret";
    #$out =~ s/\n/<br>/g;
    $out .= "\n";
    $self->output_result( $self->output_result . $out);
  }

  sub read {
    my ($self) = @_;
    my $prompt = $self->prompt;
    $self->output_result(
      $self->output_result
      . "\n"
      . $self->prompt
    );
    my $out = $self->output_result;
    my $id = $self->request->session_id;
    $self->request->print(qq{
      <div class=dialog id=repl title='Read-Eval-Print'>
        <pre>$out<input type=text name=cmd id=cmd size=60><input type=submit name=send value="Send">
        </pre>
        <script>document.getElementById('cmd').focus()</script>
      </div>
    });
    $self->yield(1);
    my $cmd = $self->request->param('cmd');
    $self->yield(0) if $cmd && $cmd eq 'exit';
    $self->output_result(
      $self->output_result
      . $cmd
      . '<br>'
    ) if $cmd;
    $cmd = undef if $cmd && $cmd eq 'exit';
    return $cmd;
  }

  method main {
    while($self->run_once) { }
  }
}

use Moose;
use Method::Signatures;

# For now we'll force Term::ReadLine::Perl since GNU doesn't work here
BEGIN { $ENV{PERL_RL} = 'Perl' }

use Devel::REPL;
use Coro;
use Coro::Event;

has repl => (is => 'rw');
has request => (is => 'rw');

=head1 METHODS

=head2 $c_repl = Continuity::REPL->new( repl => $repl );

Create and start a new REPL on the command line. Optionally pass your own Devel::REPL object. If you don't pass in $repl, a default is created.

=cut

sub BUILD {
  my $self = shift;
  unless($self->repl) {
    $self->repl( $self->default_repl );
  }
     # my $timer = Coro::Event->timer(interval => 0 );
  # async {
     # while ($timer->next) {
       # $self->repl->run_once;
     # }
  # };
  return $self;
}

=head2 default_repl

This internal method creates the default REPL if one isn't specified.

=cut

sub default_repl {
  my $self = shift;

  my $repl = Devel::REPL::Continuity->new( request => $self->request );
  $repl->load_plugin($_) for qw(
    FancyPrompt
    Packages
    Refresh
    Interrupt
    ShowClass
    History
    MultiLine::PPI
    LexEnv
    DDS
  );
    # Carp::REPL
    # DebugHelp
    # Colors
    # Completion CompletionDriver::LexEnv
    # CompletionDriver::Keywords

  $repl->fancy_prompt(sub {
    my $self = shift;
    sprintf '%s:%03d%s&gt; ',
      $self->can('current_package') ? $self->current_package : 'main',
      $self->lines_read,
      $self->can('line_depth') ? ':' . $self->line_depth : '';
  });

  $repl->fancy_continuation_prompt(sub {
    my $self = shift;
    my $pkg = $self->can('current_package') ? $self->current_package : 'main';
    $pkg =~ s/./ /g;
    sprintf '%s     %s* ',
      $pkg,
      $self->lines_read,
      $self->can('line_depth') ? $self->line_depth : '';
  });

  $repl->current_package('main');

  return $repl;
}

method process {
  return $self->repl->process();
}


=head1 SEE ALSO

L<Continuity>, L<Devel::REPL>, L<Coro::Debug>

=head1 AUTHOR

  Brock Wilcox <awwaiid@thelackthereof.org> - http://thelackthereof.org/

=head1 COPYRIGHT

  Copyright (c) 2008 Brock Wilcox <awwaiid@thelackthereof.org>. All rights
  reserved.  This program is free software; you can redistribute it and/or modify
  it under the same terms as Perl 5.10 or later.

=cut

1;



lib/CGI/Inspect.pm
package CGI::Inspect;

=head1 NAME

CGI::Inspect - Inspect and debug CGI apps with an in-browser UI

=head1 SYNOPSIS

  use CGI::Inspect; # This exports inspect()

  print "Content-type: text/html\n\n";

  my $food = "toast";

  for my $i (1..10) {
    print "$i cookies for me to eat...<br>";
    inspect() if $i == 5; # be sure to edit $toast :)
  }

  print "I also like $food!";

=head1 DESCRIPTION

This class is a drop-in web based inspector for plain CGI (or CGI-based)
applications. Include the library, and then call inspect(). In your server
error logs you'll see something like "Please connect to localhost:8080". When
you do, you'll be greeted with an inspection UI which includes a stack trace,
REPL, and other goodies.

To work it's magic this modules uses Continuity, Devel::LexAlias,
Devel::StackTrace::WithLexicals, and their prerequisites (such as PadWalker).
Remember that you can always install such things locally for debugging -- no
need to install them systemwide (in case you are afraid of the power that they
provide).

=cut

use strict;
use Continuity;
use Continuity::RequestCallbacks;
use base 'Exporter';
our @EXPORT = qw( inspect );

our $VERSION = '0.5';

=head1 EXPORTED SUBS

=head2 inspect()

This starts the Continuity server and inspector on the configured port
(defaulting to 8080). You can pass it parameters which it will then pass on to
CGI::Inspect->new. The most useful one is probably port:

  inspect( port => 12345 );

Another useful parameter is plugins. The default is:

    plugins => [qw( BasicLook Exit REPL CallStack )]

=cut

sub inspect {
  print STDERR "Starting inspector...\n";
  require IO::Handle;
  STDERR->autoflush(1);
  STDOUT->autoflush(1);
  print "<script>window.open('http://localhost:8080/','cgi-inspect');</script>\n";
  my $self = CGI::Inspect->new(@_);
  $self->start_inspecting;
}

# This might be cool, but we'll disable it for now. I mean... because it doesn't work.
# sub import {
  # my $class = shift;
  # my $nodie = grep { $_ eq 'nodie' } @_;
  # $SIG{__DIE__} = \&CGI::Inspect::inspect unless $nodie;
  # $class->SUPER::import(@_);
# }


=head1 PLUGINS

CGI::Inspect comes with a few plugins by default:

=over 4

=item * L<REPL|CGI::Inspect::Plugin::REPL> - A simple Read-Eval-Print prompt

=item * L<StackTrace|CGI::Inspect::Plugin::StackTrace> - A pretty stack trace (with lexical editing!)

=item * L<Exit|CGI::Inspect::Plugin::Exit> - Stop inspecting

=back

=head2 Creating Plugins

Plugins are easy to create! They are really just subroutines that return a
string for what they want printed. All of the built-in plugins actuall inherti
from L<CGI::Inspect::Plugin>, which just provides some convenience methods. The
main CGI::Inspect module will create an instance of your plugin with
Plugin->new, and then will execute it with $plugin->process.

Plugins can, however, make use of Continuity, including utilizing callbacks.
Here is the complete source to the 'Exit' plugin, as a fairly simple example.

  package CGI::Inspect::Plugin::Exit;

  use strict;
  use base 'CGI::Inspect::Plugin';

  sub process {
    my ($self) = @_;
    my $exit_link = $self->request->callback_link(
      Exit => sub {
        $self->manager->{do_exit} = 1;
      }
    );
    my $output = qq{
      <div class=dialog title="Exit">
        $exit_link
      </div>
    };
    return $output;
  }

  1;

For now, read the source of the default plugins for further ideas. And of
course email the author if you have any questions or ideas!

=head1 METHODS

These methods are all internal. All you have to do is call inspect().

=head2 CGI::Inspect->new()

Create a new monitor object.

=cut

sub new {
  my ($class, %params) = @_;
  my $self = {
    port => 8080,
    plugins => [qw(
      BasicLook Exit REPL CallStack
    )],
    # REPL CallStack Exit Counter FileEdit
    plugin_objects => [],
    html_headers => [],
    %params
  };
  bless $self, $class;
  return $self;
}

=head2 $self->start_inspecting

Load plugins and start inspecting!

=cut

sub start_inspecting {
  my ($self) = @_;
  $self->load_plugins;
  $self->start_server;
}
  
  # use Devel::StackTrace::WithLexicals;
  # my $trace = Devel::StackTrace::WithLexicals->new(
    # ignore_package => [qw( Devel::StackTrace CGI::Inspect )]
  # );
  # $self->trace( $trace );

=head2 $self->start_server

Initialize the Continuity server, and begin the run loop.

=cut

sub start_server {
  my ($self) = @_;
  my $docroot = $INC{'CGI/Inspect.pm'};
  $docroot =~ s/Inspect.pm/Inspect\/htdocs/;
  #print STDERR "docroot: $docroot\n";
  my $server = Continuity->new(
    port => $self->{port},
    docroot => $docroot,
    callback => sub { $self->main(@_) },
    #debug_callback => sub { STDERR->print("@_\n") },
  );
  $server->loop;
  print STDERR "Done inspecting!\n";
}

=head2 $self->display

Display the current page, based on $self->{content}

=cut

sub display {
  my ($self, $content) = @_;
  my $id = $self->request->session_id;
  my $html_headers = join '', @{ $self->{html_headers} };
  if($self->request->param('no_html_wrapper')) {
    $self->request->print($content);
  } else {
    $self->request->print(qq|
      <html>
        <head>
          <title>CGI::Inspect</title>
          $html_headers
        </head>
        <body class=smoothness>
          <form method=POST action="/">
            <input type=hidden name=sid value="$id">
            $content
          </form>
        </body>
      </html>
    |);
  }
}

=head2 $self->request

Returns the current request obj

=cut

sub request {
  my ($self) = @_;
  return $self->{request};
}

=head2 $self->load_plugins

Load all of our plugins.

=cut

sub load_plugins {
  my ($self) = @_;
  my $base = "CGI::Inspect::Plugin::";
  foreach my $plugin (@{ $self->{plugins} }) {
    my $plugin_pkg = $base . $plugin;
    eval "use $plugin_pkg";
    print STDERR "Error loading $plugin_pkg, $@\n" if $@;
    my $plugin_instance = $plugin_pkg->new( manager => $self );
    push @{ $self->{plugin_objects} }, $plugin_instance;
    $self->{plugins_by_name}->{$plugin_pkg} = $plugin_instance;
  }
}

=head2 $self->main

This is executed as the entrypoint for inspector sessions.

=cut

sub main {
  my ($self, $request) = @_;
  $self->{request} = $request; # For plugins to use
  $self->{do_exit} = 0;
  do {
    my $content = '';
    if($request->param('plugin')) {
      $content .= $self->{plugins_by_name}->{$request->param('plugin')}->process();
    } else {
      foreach my $plugin (@{$self->{plugin_objects}}) {
        $content .= $plugin->process();
      }
    }
    $self->display($content);
    $request->next->execute_callbacks
      unless $self->{do_exit};
  } until($self->{do_exit});
  $request->print("<script>window.close();</script>");
  Coro::Event::unloop();
  $request->print("Exiting...");
  $request->end_request;
}

=head1 SEE ALSO

L<Carp::REPL>

=head1 AUTHOR

  Brock Wilcox <awwaiid@thelackthereof.org> - http://thelackthereof.org/

=head1 COPYRIGHT

  Copyright (c) 2008-2009 Brock Wilcox <awwaiid@thelackthereof.org>. All rights
  reserved.  This program is free software; you can redistribute it and/or
  modify it under the same terms as Perl itself.

=cut

1;

